### PR TITLE
MAINT: Simplify  _leading_trailing

### PR DESCRIPTION
**Pull request by eric-wieser**
Removes the list comprehensions in favor of numpy primitives.
Now recurses over the indices, not the values.

Split from #10123.

### CODE REVIEW

File: numpy/core/arrayprint.py [hunk 1]
```diff
@@ -272,22 +272,24 @@ def get_printoptions():
     """
     return _format_options.copy()
 
-def _leading_trailing(a):
+
+def _leading_trailing(a, index=()):
+    """
+    Keep only the N-D corners (leading and trailing edges) of an array
+    """
     edgeitems =  _format_options['edgeitems']
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: This is a bug (it ignored the kwarg to `array2string`), but I want to leave that to another PR

File: numpy/core/arrayprint.py [hunk 2]
```diff
@@ -272,22 +272,24 @@ def get_printoptions():
...
-            b = concatenate((a[:edgeitems], a[-edgeitems:]))
-        else:
-            b = a
+    axis = len(index)
+    if axis == a.ndim:
+        return a[index]
+
+    if a.shape[axis] > 2*edgeitems:
+        return concatenate((
+            _leading_trailing(a, index + np.s_[ :edgeitems,]),
```
Comment (mhvk) [inline numpy/core/arrayprint.py]: That `,` at the end really threw me of: I'd either use `np.index_exp[:edgeitems]` (which always returns a tuple), or go for the, for me, more natural `(slice(None, edgeitems),)`

**Review by mhvk — COMMENTED** [review]
I think @ahaldane already approved this more or less, and I think it looks good too, with one suggestion.

File: numpy/core/arrayprint.py (hunk 1 above)
Comment (mhvk) [reply numpy/core/arrayprint.py]: probably good to put up a separate issue for it (if you haven't done so already) or link to a PR. (I'm only arguing from knowing my state of memory -- I'd forget...)

File: numpy/core/arrayprint.py (hunk 2 above)
Comment (eric-wieser) [reply numpy/core/arrayprint.py]: `s_` is slightly faster, but I guess we don't care. I'll switch to `index_exp`

File: numpy/core/arrayprint.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/arrayprint.py]: #10136

**Review by ahaldane — APPROVED** [review]
LGTM

**Comment by mhvk** [comment]
OK, tests passed and approved twice, so merging.
